torch/utils/data/_typing.py

```diff
@@ -338,6 +344,39 @@ def __hash__(self):
         return hash((self.__name__, self.type))
 
 
+def hook_iterator(namespace, profile_name):
+
+    def context():
+        return torch.autograd.profiler.record_function(profile_name)
+
+    func = namespace['__iter__']
+
+    if inspect.isgeneratorfunction(func):
+        @functools.wraps(func)
+        def wrap_generator(*args, **kwargs):
+            gen = func(*args, **kwargs)
+            try:
+                with context():
+                    response = gen.send(None)
+                while True:
+                    request = yield response
+                    with context():
+                        response = gen.send(request)
+            except StopIteration as e:
+                return e.value
```

Comment:
Could we make it simpler as:
```py
gen = func(*args, **kwargs)
for d in gen:
    with context():
        yield d
```

Reply:
The logic in you proposal is different with the logic of current PR. 
The existing behavior is to add context manager for the process of generating next value (the send call) instead of the yield. So, if it is the for loop, we'd like to add the context manager for the gen.__next__ call.

Reply:
Actually, only the next(gen) need to be in context manager and the yield should not be embraced with context manager. 

On another hand, the try/except is needed. Otherwise, the following similar error will be shown when it is used in for loop.

![image](https://user-images.githubusercontent.com/817030/140235965-d80cd3ee-e22b-4758-8981-80d566d10620.png)

Reply:
Makes sense. Then, let's do `try-except` and `pass` when `StopIteration` rather than returning a `e.value`.

Reply:
the behavior is consistent with existing https://github.com/pytorch/pytorch/blob/master/torch/autograd/grad_mode.py#L74.